node/node.go

```diff
@@ -264,30 +274,57 @@ func New(host host.Host, consensus *bft.Consensus, db *ethdb.LDBDatabase) *Node
 	// Setup initial state of syncing.
 	node.syncingState = NotDoingSyncing
 	node.StopPing = make(chan struct{})
+	node.Consensus.ConsensusBlock = make(chan *types.Block)
+	node.Consensus.VerifiedNewBlock = make(chan *types.Block)
+	node.peerRegistrationRecord = make(map[uint32]*syncConfig)
 
 	node.OfflinePeers = make(chan p2p.Peer)
 	go node.RemovePeersHandler()
 
 	return &node
 }
 
-// DoSyncing starts syncing.
-func (node *Node) DoSyncing() {
-	// If this node is currently doing sync, another call for syncing will be returned immediately.
-	if !atomic.CompareAndSwapUint32(&node.syncingState, NotDoingSyncing, DoingSyncing) {
-		return
+// IsOutOfSync checks whether the node is out of sync by comparing latest block with consensus block
+func (node *Node) IsOutOfSync(consensusBlock *types.Block) bool {
+	currentBlock := node.blockchain.CurrentBlock()
+	latestHash := currentBlock.Hash()
+	parentHash := consensusBlock.ParentHash()
+	consensusHash := consensusBlock.Hash()
+	// in general this case should not happen
+	if bytes.Compare(latestHash[:], consensusHash[:]) == 0 {
+		return false
 	}
-	defer atomic.StoreUint32(&node.syncingState, NotDoingSyncing)
-	if node.stateSync == nil {
-		node.stateSync = syncing.GetStateSync()
+	if bytes.Compare(latestHash[:], parentHash[:]) == 0 {
+		return false
 	}
-	if node.stateSync.StartStateSync(node.GetSyncingPeers(), node.blockchain) {
-		node.log.Debug("DoSyncing: successfully sync")
-		if node.State == NodeJoinedShard {
+	return true
+}
+
+// DoSyncing wait for check status and starts syncing if out of sync
+func (node *Node) DoSyncing() {
+	for {
+		consensusBlock := <-node.Consensus.ConsensusBlock
```

Comment:
I see this channel gets new block from consensus. But if the node is out of sync it won't be able to verify block and reach the consensus, right?

Reply:
Right, the current logic is the latest code will first check if it can verify the new block in the consensus, if not be able to verify, send consensusBlock to IsOutOfSync function.